docs: add guide for working with syntax trees in TypeScript

ChristianMurphy opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/vfile/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/vfile/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/vfile/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Avfile&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

This provides an overview of the packages around the Unified ecosystem that makes working with syntax trees easier.
Specifically focusing on:

* The basic syntax tree types
* How to include syntax tree typings for a specific language
* How to traverse a syntax tree
* How to narrow generic `Node`s to specific syntax types
* How to build a syntax tree

---

[rendered version](https://github.com/unifiedjs/unifiedjs.github.io/blob/b4423ee77650bc55db6f087d6e2db75c2dbfe112/doc/learn/syntax-trees-with-typescript.md)

---

:notebook: This does not cover how to type a `Plugin` in TypeScript or TypeScript+JSDoc, my current thought is that would be small enough and distinct enough to be recipe(s)

<!--do not edit: pr-->

ChristianMurphy commented:
/cc @remcohaszing @jounqin

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+## Working with syntax trees in TypeScript
+
+This guide will introduce you to:
```
If you add a `### Contents`, the table will be automatically generated.

ChristianMurphy replied:
good idea, added

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+  * [`hastscript`](#hastscript)
+  * [`xastscript`](#xastscript)
+
+### The Basics
```
```suggestion
### Basics
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+All unified syntax trees are based off [the **Uni**versal **S**yntax **T**ree (`unist`)](https://github.com/syntax-tree/unist).
+Unist is a types only package available on npm at [`@types/unist`](https://www.npmjs.com/package/@types/unist), and provides three interfaces which the rest of unified's syntax trees build on: `Node`, `Literal`, and `Parent`.
+
+#### Node
```
```suggestion
#### `Node`
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+}
+```
+
+#### Literal
````
```suggestion
#### `Literal`
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+}
+```
+
+#### Parent
````
```suggestion
#### `Parent`
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+
+#### Pulling unist into a project
+
+```bash
````
````suggestion
Install:

```bash
````

wooorm replied:
Same suggestions for mdast/hast/xast

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+npm install --save-dev @types/unist
+```
+
+then import the types into a TypeScript file with:
````
```suggestion
To import the types into a TypeScript file, use:
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+import type {Node, Literal, Parent} from 'unist'
+```
+
+or into a [JSDoc TypeScript](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html) project with:
````
```suggestion
To import the types into [JSDoc TypeScript](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html), use:
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+### The Languages
+
+`unist` provides the building blocks for having consistent structure, but is often more generic than what we as developers and content authors want to work with.
```
Lot of complex words. Maybe there’s a better way to describe that unist provides abstract interfaces

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+`unist` provides the building blocks for having consistent structure, but is often more generic than what we as developers and content authors want to work with.
+We want to work with a specific language like Markdown, HTML, XML, and others.
+Each of these languages has it's own syntax tree standard which extends `unist` with more specific content types.
+Let's take a look at a few of Unified's compatible languages.
```
```suggestion
Let's take a look at a few of unified's compatible languages.
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+### The Languages
+
+`unist` provides the building blocks for having consistent structure, but is often more generic than what we as developers and content authors want to work with.
+We want to work with a specific language like Markdown, HTML, XML, and others.
```
```suggestion
We want to work with a specific language like markdown, HTML, XML, and others.
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+#### MDAST (Markdown)
+
+The [**M**arkdown **A**bstract **S**yntax **T**ree (MDAST)](https://github.com/syntax-tree/mdast#readme) extends `unist` with types specific for Markdown such as [`Heading`](https://github.com/syntax-tree/mdast#heading), [`Code`](https://github.com/syntax-tree/mdast#code), [`Link`](https://github.com/syntax-tree/mdast#link), and many more.
```
```suggestion
The [**M**arkdown **A**bstract **S**yntax **T**ree (MDAST)](https://github.com/syntax-tree/mdast#readme) extends `unist` with types specific for markdown.
Some examples are [`Heading`](https://github.com/syntax-tree/mdast#heading), [`Code`](https://github.com/syntax-tree/mdast#code), and [`Link`](https://github.com/syntax-tree/mdast#link).
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+ */
+```
+
+### How to traverse a syntax tree
````
Make this section more specific to typescript. There is already an in-depth article on tree traversal. So cross link to and from that: https://github.com/unifiedjs/unifiedjs.github.io/blob/main/doc/learn/tree-traversal.md

ChristianMurphy replied:
linked to tree traversal guide

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+console.log(markdownFile.toString())
+```
+
+Or we could want on all links which are not HTTPS:
````
ungrammatical sentence

ChristianMurphy replied:
good catch, updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+#### `unist-util-visit-parents`
+
+Sometimes in addition to wanting to find a `Node` you also need to know the `Node`s higher in the tree, it's parents.
```
```suggestion
Sometimes in addition to wanting to find a `Node` you also need to know the `Node`s higher in the tree, its parents.
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+#### `unist-util-select`
+
+In some cases it can be useful to find a `Node` relative to nearby `Nodes`.
```
```suggestion
In some cases it can be useful to find a `Node` close to other `Node`s.
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,523 @@
+
+### How to narrow generic `Node` to specific syntax types
+
+Unified works with many languages, and can pull content from strings, from files, and from virtual files.
+In many cases unified doesn't know what specific `type` a `Node` will be ahead of time.
```
I think these two intro lines are superfluous backstory. I suggest removing

ChristianMurphy replied:
removed

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,523 @@
+    <releasedate>1984-04-06</releasedate>
+  </album>
+)
+```
````
Guide abruptly ends. I’m missing somethings like Conclusion. Also a Further exercises / Further reading section would be good to have

ChristianMurphy replied:
summary has been added

wooorm reviewed (commented):
Great start! I think this covers the things that TS folks will wonder about and we can link to from issues.

Some initial small suggestions inline, then the following two points:

- Running `npm install` and `npm run format` will check grammar/spelling/punctuation and the like. One example is I see a couple long, complex sentences, that that should warn about. `npm run generate` should also work, without crawling, to preview the page.

- The code examples seem to not match this projects style. I have Atom set up to format with xo and prettier on save, so what I do is to save e.g. a `.ts` extension as `example.ts`, and copy the formatted result into the markdown.

---

Finally, there are many things yelling for attention in this guide currently: casing, links, headings, a lot of inline code, a lot of fenced code. In comparison, very few things are *just plain prose*. I think it’s good to pass through the text again to highlight the important parts (for the user, in each section) with inline code/links, and lowlight some other parts!

JounQin commented:
Should we wait https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54413 first so that the guide could be better to use generic types?

wooorm commented:
@JounQin What would that change in this guide?

JounQin commented:
@wooorm Guide to use generic types for better `Node#data` type, `Literal#value` or `Parent#children`, etc.

```ts
// before
export interface StringLiteral extends Literal {
  value: string
}

// after
export interface StringLiteral extends Literal<string> {
}

// or
export type StringLiteral = Literal<string>
```

wooorm commented:
Such, or similar, behavior is not mentioned in this guide, so I don’t see a need to wait for it?

As I understand it, 54413 focusses on `unist` and making it easier to *create* things like `@types/mdast`/`@types/hast`/`@types/xast`.
This guide focusses on how to *use* `@types/mdast`/`@types/hast`/`@types/xast`.

JounQin commented:
It can also be used by end user for simple usage.

```ts
// before
declare const parent: Parent

const children = parent.children as Literal[] // error now

// after
declare const parent: Parent<Literal>

const children = parent.children // alreay `Literal[]`
```

wooorm commented:
I don‘t see any practical reasons for end users to do that.

To repeat from the other threads, I don’t think end users should do that.
They’re supposed to use mdast/hast/xast types.

wooorm commented:
On the conversation earlier between @JounQin and me: I can see some value in explaining how to type custom nodes. Especially with @remcohaszing new registry’s

ChristianMurphy commented:
> Running npm install and npm run format will check grammar/spelling/punctuation and the like. One example is I see a couple long, complex sentences, that that should warn about. npm run generate should also work, without crawling, to preview the page.

Done

> The code examples seem to not match this projects style. I have Atom set up to format with xo and prettier on save, so what I do is to save e.g. a .ts extension as example.ts, and copy the formatted result into the markdown.

I checked the examples with prettier and XO using a new file and `npm run format`, they do appear to match your preferred code style as is? :thinking: 

> Finally, there are many things yelling for attention in this guide currently: casing, links, headings, a lot of inline code, a lot of fenced code. In comparison, very few things are just plain prose. I think it’s good to pass through the text again to highlight the important parts (for the user, in each section) with inline code/links, and lowlight some other parts!

I'm confused by this.
There are too many things yelling for attention, and the solution is to add more of the things yelling for attention? :sweat_smile: 

> I can see some value in explaining how to type custom nodes. Especially with @remcohaszing new registry’s

:thinking: do you see this being a part of the basic syntax trees with TypeScript guide?
Or more being an advanced guide or a recipe?

wooorm commented:
> 🤔 do you see this being a part of the basic syntax trees with TypeScript guide?
Or more being an advanced guide or a recipe?

Maybe. At least the discussions shows me that there’s something somewhere we should explain.
While not very common, remark + micromark extensions are a thing.
And custom nodes are useful in certain cases, whether it’s remark/rehype/etc.

> There are too many things yelling for attention, and the solution is to add more of the things yelling for attention? 😅

I mean to lowlight most, and highlight less. But definitely highlight certain important parts.
e.g., I think `` `Node` `` and such could be code in certain cases, but after a while just using the word `node` is fine.

wooorm commented:
> I checked the examples with prettier and XO using a new file and npm run format, they do appear to match your preferred code style as is? 🤔

Specifically I saw the indent at 4 instead of 2 spaces?
And some double quote use, instead of single quotes

ChristianMurphy commented:
> I mean to lowlight most, and highlight less. But definitely highlight certain important parts.
> e.g., I think `Node` and such could be code in certain cases, but after a while just using the word node is fine.

Updated to lessen amount of inline `type` and `Node` highlights.

> Specifically I saw the indent at 4 instead of 2 spaces?
> And some double quote use, instead of single quotes

Updated in 0119321 and b4423ee, double quotes remain in jsx examples as prettier/xo appear to still prefer them in that context.

wooorm commented:
Nice! I’ll do another whole review tomorrow!

ChristianMurphy commented:
To @JounQin's points in https://github.com/unifiedjs/unifiedjs.github.io/pull/36#issuecomment-877590787 and https://github.com/unifiedjs/unifiedjs.github.io/pull/36#issuecomment-877592209 I see both of those potentially having value.

My intent with this first guide is focused more on people interested in adopting unified with remark.
It shows how to use `unified`, `remark`, `mdast`, etc but don't go much into how to customize them or publish a utility/module.

I could see several related topic focused more on maintaining or customizing unified, having their own recipes or guides.
Including but not limited to:
* Publishing a pure TS unified plugin
* Publishing a JSDoc based unified plugin
* Adding custom node types
* Testing plugin/syntax tree typings

ChristianMurphy commented:
Another idea, I'm not quite sure where to slot this in the guide, would be covering some recommended _tsconfig.json_ options.
In particular:
* `"module": "es6"` and `"moduleResolution": "node"` are needed to work with ESM
* `"strict": true` while not required, helps find more potential issues
* `"allowJs": true` and `"checkJs": true` are needed for TypeScript in JSDoc mode

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+## Working with syntax trees in TypeScript
+
+This guide will introduce you to:
```
```suggestion
This guide will introduce you to using unist and unified with TypeScript.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+All unified syntax trees are based off [the **Uni**versal **S**yntax **T**ree (`unist`)](https://github.com/syntax-tree/unist).
+Unist is a types only package available on npm at [`@types/unist`](https://www.npmjs.com/package/@types/unist),
+and provides three interfaces which the rest of unified’s syntax trees
+build on: `Node`, `Literal`, and `Parent`.
```
```suggestion
All unified syntax trees are based off [unist (**uni**versal **s**yntax **t**ree)](https://github.com/syntax-tree/unist).
The core types are available in a types only package: [`@types/unist`](https://www.npmjs.com/package/@types/unist).
The main type is `Node`.
Everything else extends it.
`Literal` and `Parent` are more practical types which also extend `Node`.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+kinds of content such a `Heading` or a `Link`, a `Paragraph` or a `Blockquote`
+(among others). Each of these different types extend `Node` (sometimes
+indirectly through `Literal` or `Parent`) and set `type` to a [string literal](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types)
+which uniquely identifies a kind of content (in TypeScript parlance a [discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)).
```
```suggestion
`Node` is the syntactic unit of syntax trees.
Each node extends `Node` (sometimes indirectly through `Literal` or `Parent`)
and set `type` to a [string literal](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types)
The type field tells us what kind of syntax the node is.
This field uniquely identifies a kind of content (in TypeScript parlance a
[discriminated union](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)).
For example in Markdown (mdast) `Node` will be extended to make different kinds
of content such as a `Heading` or `Link`, which respectively use a `type` field
of `'heading'` and `'link'`.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+A node also can optionally include `Data`.
+Data is a dictionary/object which can store extra information and metadata
+which is not standard to a given node type.
```
```suggestion
A node also can optionally include a `Data` interface at the `data` field.
This is an object (dictionary) that stores extra metadata which is not standard
to the node but defined by the ecosystem (utilities and plugins).
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+which is not standard to a given node type.
+
+When a syntax tree is parsed from a file, it may include `Position`, which is
+information about the Node’s location in source file.
```
```suggestion
When a syntax tree is parsed from a file, it includes positional information:
a `Position` interface at the `position` field.
This describes where the node occurred in the source file.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+#### `Literal`
+
+`Literal` extends `Node` and adds a `value` property.
+For example a markdown `text` node extends `Literal` and sets `value` to be a `string`.
```
```suggestion
For example a markdown `code` node extends `Literal` and sets `value` to be a `string`.
```

More practical example (“text” is a bit vague, “code” is I think easier to grasp)

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+*   [How to build syntax tree](#how-to-build-syntax-tree)
+*   [Summary](#summary)
+
+### The basic syntax tree types
```
Does this section really explain the basics?

When I read “the basics” I’d think: here’s the bare minimum of what I need to know to get started as a beginner with types”. This section instead (quite nicely) explains unist in depth.

Perhaps we need a new (short, and maybe we can pull some paragraphs from the rest of the document) section for the basics here. Then have this section called `## unist (universal syntax tree)`. And “unwrap” The languages:

* basics
* unist
* mdast
* hast
* xast
* summary

Another idea: change the three `How to ...` sections into their own recipes, and link to them from here? What do you think?

ChristianMurphy replied:
drafted a version of this at https://github.com/unifiedjs/unifiedjs.github.io/pull/37

wooorm replied:
Let a review there. I quite like that. Does it work well when the site is built locally too?

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+with. We want to work with a specific language like markdown, HTML, XML, and
+others. Each of these languages has it’s own syntax tree standard which extends
+`unist` with more specific content types. Let’s take a look at a few of
+unified’s compatible languages.
```
```suggestion
The types provided by unist are abstract interfaces.
In many cases, you will instead use more specific interfaces depending on what
language you’re working with.
Each language supported by unified, like markdown, HTML, and XML, has its own
syntax tree standard which extends `unist`.
Let’s take a look at a few of them.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+extends `unist` with types specific for Markdown such as [`Heading`](https://github.com/syntax-tree/mdast#heading),
+[`Code`](https://github.com/syntax-tree/mdast#code), [`Link`](https://github.com/syntax-tree/mdast#link),
+and many more. A full list of node types can be found in the [MDAST documentation](https://github.com/syntax-tree/mdast#readme).
+[Typings for MDAST are available on npm](https://www.npmjs.com/package/@types/mdast) can be installed with a package manager.
```
```suggestion
[mdast (**m**arkdown **a**bstract **s**yntax **t**ree)](https://github.com/syntax-tree/mdast#readme)
extends unist with types specific for markdown such as `Heading`, `Code`,
`Link`, and many more.
A full list of nodes can be found in the [specification](https://github.com/syntax-tree/mdast#readme).
The types are available in a types only package: [`@types/mdast`](https://www.npmjs.com/package/@types/mdast).
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+The [**H**ypertext **A**bstract **S**yntax **T**ree (HAST)](https://github.com/syntax-tree/hast#readme) extends `unist` with types specific for HTML such as [`Element`](https://github.com/syntax-tree/hast#element), [`Comment`](https://github.com/syntax-tree/hast#comment), [`DocType`](https://github.com/syntax-tree/hast#doctype), and many more.
+A full list of node types can be found in the [HAST documentation](https://github.com/syntax-tree/hast#readme).
+[Typings for HAST are available on npm](https://www.npmjs.com/package/@types/hast) can be installed with a package manager.
```
```suggestion
[hast (**h**ypertext **a**bstract **s**yntax **t**ree)](https://github.com/syntax-tree/hast#readme)
extends unist with types specific for HTML such as `Element`, `Comment`,
`DocType`, and many more.
A full list of nodes can be found in the [specification](https://github.com/syntax-tree/hast#readme).
The types are available in a types only package: [`@types/hast`](https://www.npmjs.com/package/@types/hast).
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+[`CData`](https://github.com/syntax-tree/xast#cdata), [`Instruction`](https://github.com/syntax-tree/xast#instruction),
+and many more. A full list of node types can be found in the [XAST documentation](https://github.com/syntax-tree/xast#readme).
+[Typings for XAST are available on npm](https://www.npmjs.com/package/@types/xast)
+can be installed with a package manager.
```
```suggestion
[xast (e**x**tensible **a**bstract **s**yntax **t**ree)](https://github.com/syntax-tree/xast#readme)
extends unist with types specific for HTML such as `Element`, `CData`,
`Instruction`, and many more.
A full list of nodes can be found in the [specification](https://github.com/syntax-tree/xast#readme).
The types are available in a types only package: [`@types/xast`](https://www.npmjs.com/package/@types/xast).
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+### How to traverse a syntax tree
+
+:notebook: consider reading the [introduction to tree traversal in JavaScript](./tree-traversal)
+before reading this section.
```
```suggestion
:notebook: please read the [introduction to tree traversal in JavaScript](./tree-traversal/)
before reading this section.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+often that includes finding specific types of content, then changing the
+content, validating or linting the content, or transforming the content.
+To do this we need to be able to traverse the syntax tree looking for content.
+Unified provides several type-safe utilities which can help with this.
```
```suggestion
A frequent task when working with unified is to traverse trees to find certain
nodes and then doing something with them (often validating or transforming
them).
There are several type-safe utilities provided by unified to help with this.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+[`unist-util-visit`](https://github.com/syntax-tree/unist-util-visit#readme)
+takes a syntax tree, a [`Test`](https://github.com/syntax-tree/unist-util-is#use),
+and a callback. The callback will be called for each node in the tree that
+passes the `Test`.
```
```suggestion
takes a syntax tree, a `Test`, and a callback.
The callback is called for each node in the tree that passes `Test`.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+  .use(() => (mdast: Node) => {
+    visit(
+      mdast,
+      // checks that the Node is a heading
```
```suggestion
      // Check that the Node is a heading:
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+  .use(() => (mdast: Node) => {
+    visit(
+      mdast,
+      // this checks both that the Node is a list and that it is ordered
```
```suggestion
      // Check that the Node is a list and that it is ordered:
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+    visit(
+      mdast,
+      // this both checks the node is a link and checks the content of url
+      (node: Node): node is Link => is<Link>(node, 'link') && !node.url.includes('https'),
```
Hmm, I think we either need a different and better example, or fix the code:
- What about relative links, anchors, other protocols?
- This allows: `./https`

Depending on what the code should do, I would go with:

```js
try {
  url = new URL(node.url, base)
  if (url.protocol === 'http:') {
    // warn
  } else {
    // fine
  }
} catch {
  // fine
}
```

ChristianMurphy replied:
I removed this example for now, this type of test is also covered by the `unist-util-is` example

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+Sometimes besides to wanting to find a node you also need to know the
+Node’s higher in the tree, its parents. [`unist-util-visit-parents`](https://github.com/syntax-tree/unist-util-visit-parents)
+is like `unist-util-visit`, but also includes a list of parent nodes.
```
```suggestion
Sometimes it’s needed to know the ancestors of a node (all its parents).
[`unist-util-visit-parents`](https://github.com/syntax-tree/unist-util-visit-parents)
is like `unist-util-visit` but includes a list of all parent nodes.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+[`unist-util-select`](https://github.com/syntax-tree/unist-util-select) lets
+us use [CSS selectors](https://github.com/syntax-tree/unist-util-select#support)
+to find `Node`s in a syntax tree. For example in markdown if we want to find
+all the `Paragraph`s inside `Blockquote`s, we could:
```
```suggestion
Sometimes CSS selectors are easier to read than several (nested) if/else
statements.
[`unist-util-select`](https://github.com/syntax-tree/unist-util-select) lets
you do just that.
For example if we want to find all `Paragraph`s that are somewhere in a
`Blockquote`, we could:
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+the type, taking the more general `Node` and doing type safe checks to get to a
+more specific type like a `Link`. Unified provides several utilities to help
+with this, and there are some TypeScript language features which can also help.
+Let’s take a look at `unist-util-is`.
```
```suggestion
To work with a specific node type or a set of node types we need to
[narrow](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) their
type.
For example, we can take a `Node` and perform a type safe check to get a more
specific type like a `Link`.
Unified provides a utility to help with this and there are some TypeScript
language features which can also help.
Let’s first take a look at `unist-util-is`.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+and returns `true` if the test passes, and `false` if it does not. It also is
+a [TypeScript predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)
+meaning if used as the condition for an `if` statement, TypeScript knows more
+about the type inside the `if`.
```
```suggestion
[`unist-util-is`](https://github.com/syntax-tree/unist-util-is#readme) takes a
`Node` and a `Test` and returns whether the test passes.
It can be used as a [TypeScript type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)
which when used as a condition (such as in an if-statement) tells TypeScript
to narrow a node.
```

ChristianMurphy replied:
updated

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+const node: Node = {type: 'example'}
+
+if (is<List>(node, 'list')) {
+  // If we get here, node is List
```
```suggestion
  // If we’re here, node is List.
  // JavaScript knows this because of the test: 'list'.
  // TypeScript knows this because of `is<List>`
  // xxx
```

I think TS beginners (including me earlier) have a hard time understand what’s doing what.
In the `xxx` part I think we should explain that `unist-util-is`’s types are set up in such a way to make sure that `List` and `'list'` match and you’ll get a TS error otherwise. (but I’m not sure how to explain it best)

ChristianMurphy replied:
added some commentary on how type matching works

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+}
+
+if (is<Heading>(node, {type: 'heading', depth: 1})) {
+  // If we get here, node is Heading
```
```suggestion
  // If we get here, node is Heading. JS will know it also has a `depth: 1`, but TS won’t.
```

ChristianMurphy replied:
I'm not sure I follow this comment. JSDoc+TS will know?
Or do you mean the value is 1 in JavaScript?

ChristianMurphy replied:
added some commentary on how type matching works

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+When content needs to be created or added, it’s often useful to build new
+syntax trees, or fragments of syntax trees. This can be easy to do with plain
+JSON, unified also offers some utilities for building trees with hyperscript
+or JSX.
```
```suggestion
It’s often useful to build new (fragments of) syntax trees when adding or
replacing content.
It’s possible to create trees with plain object and array literals (JSON) or
programmatically with a small utility.
Finally it’s even possible to use JSX to build trees.
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+#### JSON
+
+Often a tree can be created with plain JSON, for example:
```
```suggestion
The most basic way to create a tree is with plain object and arrays, such as:
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,572 @@
+    }
+  ]
+}
+```
````
I think the above two examples can be merged: it’s a TS guide, folks will want to use types

ChristianMurphy replied:
merged

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,572 @@
+}
+```
+
+#### `unist-util-builder`
````
```suggestion
#### `unist-builder`
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+For more concise, hyperscript (or [`React.createElement`](https://reactjs.org/docs/react-api.html#createelement))
+like syntax, [`unist-builder`](https://github.com/syntax-tree/unist-builder#readme)
+can be used:
```
```suggestion
It’s also possible to build trees with [`unist-builder`](https://github.com/syntax-tree/unist-builder#readme).
It allows a more concise, hyperscript (similar to `React.createElement`) like
syntax:
```

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,572 @@
+    <input type="submit" name="send" />
+  </form>
+)
+```
````
````suggestion
When working with hast (HTML), [`hastscript`](https://github.com/syntax-tree/hastscript#readme)
can be used.

```ts
import {h} from 'hastscript'

console.log(
  h('div#some-id.foo', [
    h('span', 'some text'),
    h('input', {type: 'text', value: 'foo'}),
    h('a.alpha.bravo.charlie', {download: true}, 'delta')
  ])
)
```

hastscript can also be used as a JSX pragma:

```tsx
/** @jsx h @jsxFrag null */
import {h} from 'hastscript'

console.log(
  <form method="POST">
    <input type="text" name="foo" />
    <input type="text" name="bar" />
    <input type="submit" name="send" />
  </form>
)
```
````

- I believe TS doesn’t yet support the automatic runtime?
- JSX doesn’t allow multiple syntax trees / pragmas per file
- It’s not familiar to everyone, only showing the JSX example will lead some folks to thing `h` can’t be used without it

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
````diff
@@ -0,0 +1,572 @@
+    <releasedate>1984-04-06</releasedate>
+  </album>
+)
+```
````
````suggestion
When working with xast (XML), [`xastscript`](https://github.com/syntax-tree/xastscript#readme)
can be used.

```ts
import {x} from 'xastscript'

console.log(
  x('album', {id: 123}, [
    x('name', 'Exile in Guyville'),
    x('artist', 'Liz Phair'),
    x('releasedate', '1993-06-22')
  ])
)
```

xastscript can also be used as a JSX pragma:

```tsx
/** @jsx x @jsxFrag null */
import {x} from 'xastscript'

console.log(
  <album id={123}>
    <name>Born in the U.S.A.</name>
    <artist>Bruce Springsteen</artist>
    <releasedate>1984-04-06</releasedate>
  </album>
)
```
````

wooorm commented on doc/learn/syntax-trees-with-typescript.md:
```diff
@@ -0,0 +1,572 @@
+
+### Summary
+
+*   Using TypeScript can make finding typos and bugs easier
```
I don‘t think this point is made in the guide

ChristianMurphy replied:
removed

wooorm reviewed (commented):
Nice!!!

Alright, I did a pass of the whole document.
I do think my suggestions are useful compared to the current text, but they’re perfect: please take them and amend them where appropriate

ChristianMurphy commented:
https://github.com/unifiedjs/unifiedjs.github.io/pull/36#discussion_r669035471, closing this in favor of #37
